Initial implementation of QIR Runtime (#396)
microsoft/qsharp-runtime: src/QirRuntime/README.md

This project contains headers for the native runtime, [possibly incomplete] implementations of a few
simulators, and the QIR-to-native bridge to support running QIR files against simulators from this or other native runtime(s).

- `public` folder contains the public headers
- `lib` folder contains the implementation of the runtime, the simulators and the QIR bridge
- `test` folder contains tests for the runtime
- `experimental` folder contains features for which specifications and design are in too early stage to include into the QDK.

## Build

The project is using CMake (3.17) + Ninja(1.10.0) + Clang++(10.0.0). Other versions of the tools might work but haven't been tested. Only x64 architecture is supported.

You can use CMake directly. For example, to produce Release build:

1. navigate into the project's root folder
2. mkdir build
3. cd build
4. cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ..
5. cmake --build .

Or you can run `build.py` script from the root of the project. The default options for the script are `make debug`.

- (Windows) `python build.py [make/nomake] [debug|release] [ir]`
- (Linux) `python3 build.py [make/nomake] [debug|release] [ir]`

> The scrip will place the build artefacts into `build/[Windows|Linux]/[Debug|Release]` folder.

Reviewer: ```suggestion
The script will place the build artifacts into `build/[Windows|Linux]/[Debug|Release]` folder.
```

> 

Author: The project builds on Mac and all tests pass, including those that run QIR on the full state simulator.

### Windows pre-reqs

> 1. Install Clang, Ninja and CMake from the public distros.

Author: Those are public distros we don't maintain so their location might change. User can easily discover them online (the user might also have their preferred packaging apps, etc.)

2. Add all three to your/system `%PATH%`.

> 3. Install VS 2019 (Clang uses MSVC's standard library on Windows).

Author: The native simulator requires VS's toolchain as well (it's built with msvc compiler) so this is not a new requirement for working with the repo as a whole. I don't know whether it's possible to install the toolchain without installing VS (from a quick search online, I believe the answer is "no").

> 4. Install .NET

Author: Clarified: ".NET Core 3.1". As with other prerequisites I don't want to maintain the links to public distros as those are external to us and might change.

4. <_optional_>
    1. Either clone LLVM's repo and build LLVM locally to produce your own `opt.exe` and other IR tools or download LLVM's dev package.

> 5. Install clang-tidy and clang-format if your Clang/LLVM packages didn't include the tools.

Author: The releases of llvm from https://releases.llvm.org/ include both of these so the user, who does the mainstream setup, shouldn't need to do anything here. And if the user is sophisticated enough to install LLVM/Clang from somewhere else, I think it's OK to expect them to figure out the additional tools on their own.

2. In WSL:
    1. `$ sudo apt install cmake` (`$ cmake --version` should return 3.16.3)
    2. `$ sudo apt-get install ninja-build` (`$ ninja --version` should return 1.10.0)
    3. `$ sudo apt install clang` (`$ clang++ --version` should return 10.0.0)
    4. Set Clang as the preferred C/C++ compiler:
        - $ export CC=/usr/bin/clang
        - $ export CXX=/usr/bin/clang++
    5. `$ sudo apt install clang-tidy` (`$ clang-tidy --version` should return 'LLVM version 10.0.0')
    6. <_optional_> $ sudo apt install llvm
    7. <_optional_> To use build/test scripts, check that you have python3 installed (it should be by default).

See [https://code.visualstudio.com/docs/remote/wsl] on how to use VS Code with WSL.

### Generating IR

To generate IR of the native components as part of the build:

1. Install LLVM (see above) so you have opt.exe and other tools that work with IR.
2. Modify ProduceIR.cmake to point to these tools. (Or copy them into a separate folder and add that folder to `%PATH%.` If you are building LLVM locally, don't add the whole target build folder to `%PATH%` because it would likely confuse Clang.)

> 3. To produce IR once pass `ir` option to `build.py`. To do that on every build, uncomment setting `${GENERATE_IR}` variable in the root `CMakeLists.txt`.

Author: Clang. This is the IR from our native code. I thought we would need it for optimizations and occasionally I used it to see what Clang does with native code when developing the bridge. We might remove this later but it was such a pain to enable it that I'd rather keep it in until dead sure it's not being used

All test binaries and their dependencies are copied by the build into *install* folder: `build/[Windows|Linux]/[Debug|Release]/bin` and should be run from there (otherwise the tests might fail to load the shared libraries they depend on). On **Linux** `test.py` adds the install folder to LD_LIBRARY_PATH for the duration of the script. If you'd like to run the tests directly, add the path for the session manually. On WSL it might look like this: `$export LD_LIBRARY_PATH=/mnt/d/repos/llvm-project/qtracer/build/Linux/Debug/bin:${LD_LIBRARY_PATH}`.

The project is using catch2 for all native tests, including QIR. `<test_binary> -help` provides details on how to run a subset of the tests and other options.

All native tests are fully integrated with CTest. The coverage is the same as when using `test.py` or running the test binaries individually, but the results are logged into the corresponding `build/[Windows|Linux]/[Debug|Release]/bin/<test_binary_name>_results.xml` file. To trigger tests this way, navigate into `build/[Windows|Linux]/[Debug|Release]` folder and run `ctest`. No configuration options required.

## QIR Bridge and Runtime

> This project contains implementation of the QIR runtime per the specifications at [https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR] and the translation layer between the QIR and the IR, generated by Clang from the native code. We call the translation layer "QIR Bridge". This project also provides an optional implementation of intrinsics from `__quantum__qis*` namespace that are used in QIR, generated from Q#.

Reviewer 2: ```suggestion
This project contains an implementation of the QIR runtime per the specifications at [https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR] and the translation layer between the QIR and the IR, generated by Clang from the native code. We call the translation layer "QIR Bridge". This project also provides an optional implementation of intrinsics from `__quantum__qis*` namespace that are used in QIR, generated from Q#.
```

![QIR Bridge architecture diagram](qir.png?raw=true "QIR Bridge architecture diagram")

There are two ways to compile and run the QIR files against the runtime.

1. Link against the runtime libraries *statically*. For the example of this approach see `test/QIR-static` tests. It allows the client to access the target simulator directly, if so desired.
1. Link against the *shared qdk* library. The example of this approach can be found in `test/QIR-dynamic` folder. In the future we'll provide fully self-contained packages of the runtime to enable this workflow completely outside of the current project.

> QIR's architecture assumes a single target, whether that be hardware or a particular simulator. As a result, there is no provision in the QIR specifications to choose a target dinamically. To connect QIR to the simulators from this runtime, we provide `SetCurrentQuantumApiForQIR(IQuantumApi*)` method. When linking against shared qdk, the method will be invoked automatically when creating an execution context. If linking statically, the client can call the method directly with the simulator of their choice. Switching simulators while executing QIR isn't supported, however, and would yield undefined behaviour.

Reviewer 2: ```suggestion
QIR's architecture assumes a single target, whether that be hardware or a particular simulator. As a result, there is no provision in the QIR specifications to choose a target dinamically. To connect QIR to the simulators from this runtime, we provide `SetCurrentQuantumApiForQIR(IQuantumApi*)` method. When linking against the shared qdk library, the method will be invoked automatically when creating an execution context. If linking statically, the client can call the method directly with the simulator of their choice. Switching simulators while executing QIR isn't supported, however, and would yield undefined behavior.
```

### Building from IR files
CMake doesn't support using LLVM's IR files as input so instead we invoke Clang directly from custom commands to create utility libs that can be linked into other targets using their absolute paths.

*NB*: Compiling from IR has fewer checks than compiling from C++. For example, IR doesn't support overloading so declarations

> and definitions of functions are matched by name, without taking in account the arguments. This means that a build might

Reviewer 2: ```suggestion
and definitions of functions are matched by name, without taking into account the arguments. This means that a build might
```

succeed with mismatched signatures between caller/callee which will likely lead to crashes and other bugs at runtime.

> **QIR runtime is work in progress. Current known limitations are as follows:**

Reviewer 2: ```suggestion
**The QIR runtime is work in progress. Current known limitations are as follows:**
```

1. All functionality related to BigInt type (including `__quantum__rt__bigint_to_string`) NYI.
2. QIR is assumed to be single threaded. No effort was made to make the bridge and runtime thread safe.
3. Strings are implemented as a thin wrapper over std::string with virtually no optimizations.
4. `__quantum__rt__string_create` currently doesn't conform to the spec (it expects a null terminated string rather than a string of specified length).

> 5. Variadic functions (e.g. `__quantum__rt__array_create`) require platform specific bridges. Currently implemented bridge is for Windows.

Reviewer 2: ```suggestion
5. Variadic functions (e.g. `__quantum__rt__array_create`) require platform specific bridges. The currently implemented bridge is for Windows.
```
